File: src/main/kotlin/app/softwork/hyparquet/Convert.kt
package app.softwork.hyparquet

import java.util.Date

/**
 * Default type parsers when no custom ones are given
 */
val DEFAULT_PARSERS: ParquetParsers = object : ParquetParsers {
    override fun timestampFromMilliseconds(millis: Long): Any {
        return Date(millis)
    }

    override fun timestampFromMicroseconds(micros: Long): Any {
        return Date(micros / 1000)
    }

    override fun timestampFromNanoseconds(nanos: Long): Any {
        return Date(nanos / 1000000)
    }

    override fun dateFromDays(days: Int): Any {
        return Date(days * 86400000L)
    }

    override fun stringFromBytes(bytes: ByteArray): String {
        return bytes.decodeToString()
    }
}

/**
 * Convert known types from primitive to rich, and dereference dictionary.
 *
 * @param data series of primitive types
 * @param dictionary
 * @param encoding
 * @param columnDecoder
 * @returns series of rich types
 */
fun convertWithDictionary(
    data: DecodedArray,
    dictionary: DecodedArray?,
    encoding: Encoding,
    columnDecoder: ColumnDecoder
): DecodedArray {
    if (dictionary != null && (encoding == Encoding.PLAIN_DICTIONARY || encoding == Encoding.RLE_DICTIONARY)) {
        // Dereference dictionary
        return when (data) {
            is DecodedArray.IntArrayType -> {
                val dictList = when (dictionary) {
                    is DecodedArray.AnyArrayType -> dictionary.array
                    else -> error("Unsupported dictionary type")
                }
                DecodedArray.AnyArrayType(data.array.map { dictList[it] })
            }
            is DecodedArray.ByteArrayType -> {
                val dictList = when (dictionary) {
                    is DecodedArray.AnyArrayType -> dictionary.array
                    else -> error("Unsupported dictionary type")
                }
                DecodedArray.AnyArrayType(data.array.map { dictList[it.toInt() and 0xFF] })
            }
            else -> convert(data, columnDecoder)
        }
    } else {
        return convert(data, columnDecoder)
    }
}

/**
 * Convert known types from primitive to rich.
 *
 * @param data series of primitive types
 * @param columnDecoder - uses element, utf8, and parsers properties
 * @returns series of rich types
 */
fun convert(data: DecodedArray, columnDecoder: ColumnDecoder): DecodedArray {
    val element = columnDecoder.element
    val parsers = columnDecoder.parsers ?: DEFAULT_PARSERS
    val utf8 = columnDecoder.utf8 ?: true
    val type = element.type
    val ctype = element.converted_type
    val ltype = element.logical_type

    if (ctype == ConvertedType.DECIMAL) {
        val scale = element.scale ?: 0
        val factor = Math.pow(10.0, -scale.toDouble())
        return when (data) {
            is DecodedArray.AnyArrayType -> {
                DecodedArray.AnyArrayType(data.array.map { value ->
                    when (value) {
                        is ByteArray -> parseDecimal(value) * factor
                        is Number -> value.toDouble() * factor
                        else -> value
                    }
                })
            }
            is DecodedArray.IntArrayType -> {
                DecodedArray.AnyArrayType(data.array.map { it * factor })
            }
            is DecodedArray.LongArrayType -> {
                DecodedArray.AnyArrayType(data.array.map { it * factor })
            }
            else -> data
        }
    }
    
    if (ctype == null && type == ParquetType.INT96) {
        return when (data) {
            is DecodedArray.AnyArrayType -> {
                DecodedArray.AnyArrayType(data.array.map { value ->
                    when (value) {
                        is Long -> parsers.timestampFromNanoseconds(parseInt96Nanos(value))
                        else -> value
                    }
                })
            }
            else -> data
        }
    }

    if (ctype == ConvertedType.DATE) {
        return when (data) {
            is DecodedArray.IntArrayType -> DecodedArray.AnyArrayType(data.array.map { parsers.dateFromDays(it) })
            is DecodedArray.AnyArrayType -> DecodedArray.AnyArrayType(data.array.map { 
                when (it) {
                    is Int -> parsers.dateFromDays(it)
                    else -> it
                }
            })
            else -> data
        }
    }

    if (ctype == ConvertedType.TIMESTAMP_MILLIS) {
        return when (data) {
            is DecodedArray.LongArrayType -> DecodedArray.AnyArrayType(data.array.map { parsers.timestampFromMilliseconds(it) })
            else -> data
        }
    }

    if (ctype == ConvertedType.TIMESTAMP_MICROS) {
        return when (data) {
            is DecodedArray.LongArrayType -> DecodedArray.AnyArrayType(data.array.map { parsers.timestampFromMicroseconds(it) })
            else -> data
        }
    }

    if (ctype == ConvertedType.JSON) {
        return when (data) {
            is DecodedArray.AnyArrayType -> {
                DecodedArray.AnyArrayType(data.array.map { value ->
                    when (value) {
                        is ByteArray -> kotlinx.serialization.json.Json.parseToJsonElement(value.decodeToString())
                        else -> value
                    }
                })
            }
            else -> data
        }
    }

    if (ctype == ConvertedType.BSON) {
        throw Error("parquet bson not supported")
    }

    if (ctype == ConvertedType.INTERVAL) {
        throw Error("parquet interval not supported")
    }

    if (ctype == ConvertedType.UTF8 || (ltype is LogicalType.STRING) || (utf8 && type == ParquetType.BYTE_ARRAY)) {
        return when (data) {
            is DecodedArray.AnyArrayType -> {
                DecodedArray.AnyArrayType(data.array.map { value ->
                    when (value) {
                        is ByteArray -> parsers.stringFromBytes(value)
                        else -> value
                    }
                })
            }
            else -> data
        }
    }

    if (ctype == ConvertedType.UINT_64 || (ltype is LogicalType.INTEGER && ltype.bitWidth == 64 && !ltype.isSigned)) {
        return when (data) {
            is DecodedArray.LongArrayType -> data // Keep as Long, Kotlin doesn't have unsigned arrays in same way
            else -> data
        }
    }

    if (ctype == ConvertedType.UINT_32 || (ltype is LogicalType.INTEGER && ltype.bitWidth == 32 && !ltype.isSigned)) {
        return data // Keep as is
    }

    if (ltype is LogicalType.FLOAT16) {
        return when (data) {
            is DecodedArray.AnyArrayType -> {
                DecodedArray.AnyArrayType(data.array.map { value ->
                    when (value) {
                        is ByteArray -> parseFloat16(value)
                        else -> value
                    }
                })
            }
            else -> data
        }
    }

    if (ltype is LogicalType.TIMESTAMP) {
        val parser = when (ltype.unit) {
            TimeUnit.MILLIS -> parsers::timestampFromMilliseconds
            TimeUnit.MICROS -> parsers::timestampFromMicroseconds
            TimeUnit.NANOS -> parsers::timestampFromNanoseconds
        }
        return when (data) {
            is DecodedArray.LongArrayType -> DecodedArray.AnyArrayType(data.array.map { parser(it) })
            else -> data
        }
    }

    return data
}

/**
 * @param bytes
 * @returns
 */
fun parseDecimal(bytes: ByteArray): Double {
    if (bytes.isEmpty()) return 0.0

    var value = 0L
    for (byte in bytes) {
        value = value * 256 + (byte.toLong() and 0xFF)
    }

    // handle signed
    val bits = bytes.size * 8
    if (value >= (1L shl (bits - 1))) {
        value -= (1L shl bits)
    }

    return value.toDouble()
}

/**
 * Converts INT96 date format (hi 32bit days, lo 64bit nanos) to nanos since epoch
 * @param value
 * @returns
 */
fun parseInt96Nanos(value: Long): Long {
    val days = (value shr 64) - 2440588L
    val nano = value and -1L // all bits set = 0xffffffffffffffffL
    return days * 86400000000000L + nano
}

/**
 * @param bytes
 * @returns
 */
fun parseFloat16(bytes: ByteArray?): Double? {
    if (bytes == null) return null
    val int16 = ((bytes[1].toInt() and 0xFF) shl 8) or (bytes[0].toInt() and 0xFF)
    val sign = if ((int16 shr 15) != 0) -1.0 else 1.0
    val exp = (int16 shr 10) and 0x1f
    val frac = int16 and 0x3ff
    if (exp == 0) return sign * Math.pow(2.0, -14.0) * (frac / 1024.0) // subnormals
    if (exp == 0x1f) return if (frac != 0) Double.NaN else sign * Double.POSITIVE_INFINITY
    return sign * Math.pow(2.0, (exp - 15).toDouble()) * (1 + frac / 1024.0)
}

File: src/convert.ts
/**
 * @import {ColumnDecoder, DecodedArray, Encoding, ParquetParsers} from '../src/types.js'
 */
import {ColumnDecoder, ConvertedType, DecodedArray, Encoding, ParquetParsers, ParquetType, TimeUnit} from "./types.js";

const decoder = new TextDecoder()

/**
 * Default type parsers when no custom ones are given
 */
export const DEFAULT_PARSERS: ParquetParsers = {
  timestampFromMilliseconds(millis) {
    return new Date(Number(millis))
  },
  timestampFromMicroseconds(micros) {
    return new Date(Number(micros / 1000n))
  },
  timestampFromNanoseconds(nanos) {
    return new Date(Number(nanos / 1000000n))
  },
  dateFromDays(days) {
    return new Date(days * 86400000)
  },
  stringFromBytes(bytes) {
    return bytes && decoder.decode(bytes)
  },
}

/**
 * Convert known types from primitive to rich, and dereference dictionary.
 *
 * @param {DecodedArray} data series of primitive types
 * @param {DecodedArray | undefined} dictionary
 * @param {Encoding} encoding
 * @param {ColumnDecoder} columnDecoder
 * @returns {DecodedArray} series of rich types
 */
export function convertWithDictionary(data: DecodedArray, dictionary: DecodedArray | undefined, encoding: Encoding, columnDecoder: ColumnDecoder): DecodedArray {
  if (dictionary && (encoding === Encoding.PLAIN_DICTIONARY || Encoding.RLE_DICTIONARY)) {
    let output = data
    if (data instanceof Uint8Array && !(dictionary instanceof Uint8Array)) {
      // @ts-expect-error upgrade data to match dictionary type with fancy constructor
      output = new dictionary.constructor(data.length)
    }
    for (let i = 0; i < data.length; i++) {
      output[i] = dictionary[data[i]]
    }
    return output
  } else {
    return convert(data, columnDecoder)
  }
}

/**
 * Convert known types from primitive to rich.
 *
 * @param {DecodedArray} data series of primitive types
 * @param {Pick<ColumnDecoder, "element" | "utf8" | "parsers">} columnDecoder
 * @returns {DecodedArray} series of rich types
 */
export function convert(data: DecodedArray, columnDecoder: ColumnDecoder): DecodedArray {
  const { element, parsers, utf8 = true } = columnDecoder
  const { type, converted_type: ctype, logical_type: ltype } = element
  if (ctype === ConvertedType.DECIMAL) {
    const scale = element.scale || 0
    const factor = 10 ** -scale
    const arr = new Array(data.length)
    for (let i = 0; i < arr.length; i++) {
      if (data[i] instanceof Uint8Array) {
        arr[i] = parseDecimal(data[i]) * factor
      } else {
        arr[i] = Number(data[i]) * factor
      }
    }
    return arr
  }
  if (!ctype && type === ParquetType.INT96) {
    return Array.from(data).map(v => parsers.timestampFromNanoseconds(parseInt96Nanos(v)))
  }
  if (ctype === ConvertedType.DATE) {
    return Array.from(data).map(v => parsers.dateFromDays(v))
  }
  if (ctype === ConvertedType.TIMESTAMP_MILLIS) {
    return Array.from(data).map(v => parsers.timestampFromMilliseconds(v))
  }
  if (ctype === ConvertedType.TIMESTAMP_MICROS) {
    return Array.from(data).map(v => parsers.timestampFromMicroseconds(v))
  }
  if (ctype === ConvertedType.JSON) {
    return data.map(v => JSON.parse(decoder.decode(v)))
  }
  if (ctype === ConvertedType.BSON) {
    throw new Error('parquet bson not supported')
  }
  if (ctype === ConvertedType.INTERVAL) {
    throw new Error('parquet interval not supported')
  }
  if (ctype === ConvertedType.UTF8 || ltype?.type === 'STRING' || utf8 && type === ParquetType.BYTE_ARRAY) {
    return data.map(v => parsers.stringFromBytes(v))
  }
  if (ctype === ConvertedType.UINT_64 || ltype?.type === 'INTEGER' && ltype.bitWidth === 64 && !ltype.isSigned) {
    if (data instanceof BigInt64Array) {
      return new BigUint64Array(data.buffer, data.byteOffset, data.length)
    }
    const arr = new BigUint64Array(data.length)
    for (let i = 0; i < arr.length; i++) arr[i] = BigInt(data[i])
    return arr
  }
  if (ctype === ConvertedType.UINT_32 || ltype?.type === 'INTEGER' && ltype.bitWidth === 32 && !ltype.isSigned) {
    if (data instanceof Int32Array) {
      return new Uint32Array(data.buffer, data.byteOffset, data.length)
    }
    const arr = new Uint32Array(data.length)
    for (let i = 0; i < arr.length; i++) arr[i] = data[i]
    return arr
  }
  if (ltype?.type === 'FLOAT16') {
    return Array.from(data).map(parseFloat16)
  }
  if (ltype?.type === 'TIMESTAMP') {
    const { unit } = ltype
    let parser: ParquetParsers[keyof ParquetParsers] = parsers.timestampFromMilliseconds
    if (unit === TimeUnit.MICROS) parser = parsers.timestampFromMicroseconds
    if (unit === TimeUnit.NANOS) parser = parsers.timestampFromNanoseconds
    const arr = new Array(data.length)
    for (let i = 0; i < arr.length; i++) {
      arr[i] = parser(data[i])
    }
    return arr
  }
  return data
}

/**
 * @param {Uint8Array} bytes
 * @returns {number}
 */
export function parseDecimal(bytes: Uint8Array): number {
  if (!bytes.length) return 0

  let value = 0n
  for (const byte of bytes) {
    value = value * 256n + BigInt(byte)
  }

  // handle signed
  const bits = bytes.length * 8
  if (value >= 2n ** BigInt(bits - 1)) {
    value -= 2n ** BigInt(bits)
  }

  return Number(value)
}

/**
 * Converts INT96 date format (hi 32bit days, lo 64bit nanos) to nanos since epoch
 * @param {bigint} value
 * @returns {bigint}
 */
function parseInt96Nanos(value: bigint): bigint {
  const days = (value >> 64n) - 2440588n
  const nano = value & 0xffffffffffffffffn
  return days * 86400000000000n + nano
}

/**
 * @param {Uint8Array | undefined} bytes
 * @returns {number | undefined}
 */
export function parseFloat16(bytes: Uint8Array | undefined): number | undefined {
  if (!bytes) return undefined
  const int16 = bytes[1] << 8 | bytes[0]
  const sign = int16 >> 15 ? -1 : 1
  const exp = int16 >> 10 & 0x1f
  const frac = int16 & 0x3ff
  if (exp === 0) return sign * 2 ** -14 * (frac / 1024) // subnormals
  if (exp === 0x1f) return frac ? NaN : sign * Infinity
  return sign * 2 ** (exp - 15) * (1 + frac / 1024)
}
